### ISSUE TITLE
Publish to PyPI

### ISSUE THREAD

**Issue by webknjaz** [tlsfuzzer/tlsfuzzer#903]
Hello, it doesn't seem like anybody brought this up in the future, but would you be open in publishing this to PyPI so that after a simple `pip install`, the scripts would be invokeable? I'll be happy to contribute packaging and GHA automations for this as well as looking into unifying the CLI required for this to function.
This would allow people to use it, without having to jump through the hoops of executing several commands to set up the environment in an out-of-standard manner.
My motivation is that eventually it'd be nice to be able to integrate this into testing of web frameworks (I maintain CherryPy, Cheroot and aiohttp, for example).

**Comment by tomato42** [comment]
I'm not against the concept, the issue is that we have quite a few tools that need to be accessible after installation (combine.py, extract.py, analysis.py,...) and _a lot_ of test scripts. Installing them into general PATH could easily be considered "polluting". while having simple `tlsfuzzer combine` as a wrapper that calls into `combine.py` would be rather easy to do, 95% of the worth of the projects is what's in `scripts/`.

How to make execution _and discovery_ of those (you know, with tab-completion) easy is not something I have any experience with. Combine it with basically all of my focus for the past 4 years going towards [Marvin Attack](https://people.redhat.com/~hkario/marvin/) and yeah, such niceties are not stuff I end up working on.

That being said, if you have an idea how to solve all of it in a general way (one that won't require any, or at most, very little changes to scripts) I'm fully open to discussing it, and once we agree on solution reviewing the PR and getting it merged.

**Comment by webknjaz** [comment]
So there are several ways this can be interfaced:
- exposing scripts through runpy: it's invocable via `python -m tslfuzzer.combine`
- exposing them as console-scripts prefixed with `tlsfuzzer-` would make for invocations like `tlsfuzzer-combine`
- and finally, one console-script `tlsfuzzer` with subcommands — so `tslfuzzer combine` would be the interface

All of above can be implemented and combined, of course. And it wouldn't pollute `$PATH` since everything would be auto-completed the same way. And in all cases, using any library for argument parsing would work. The second way should work exactly the same as it is now in terms of the arg completion.

As for the argparsing/CLI libs, the modern stdlib way is using argparse. And known third parties are Click and Typer. The Python 2.6 requirement has its limitations, though.

**Comment by tomato42** [comment]
ok, but this solves the issue of tools, it doesn't solve the issue for scripts.

yes, py2.6 is a limitation, and I'd rather not introduce additional dependencies: as the project is now, it and all dependencies can be put in a single tarball that requires just the python executable to execute

**Comment by webknjaz** [comment]
> ok, but this solves the issue of tools, it doesn't solve the issue for scripts.

What do you mean? I thought I was talking about the scripts.

**Comment by webknjaz** [comment]
> yes, py2.6 is a limitation, and I'd rather not introduce additional dependencies: as the project is now, it and all dependencies can be put in a single tarball that requires just the python executable to execute

What do you mean by that? Pure-python deps?

**Comment by tomato42** [comment]
> What do you mean? I thought I was talking about the scripts.

by "scripts" I mean the files that live in `scripts/` directory, they can't be loaded as modules, they have to be executed as standalone applications

by "tools" I mean the specific modules in the `tlsfuzzer/` directory that can be loaded either as a module, or executed as standalone applications. Those should already work with runpy, so stuff like `python -m tlsfuzzer.extract --help` should work.

> What do you mean by that? Pure-python deps?

for python-ecdsa and tlslite-ng (the only mandatory dependencies) to work, it's enough to do a symlink to the `ecdsa` and `tlslite` directory respectively from the tlsfuzzer repo checkout, then nothing needs to be installed on the system to run the scripts (with the exception of timing scripts). When working with old python releases (2.6, 2.7), that's a really convenient feature to have as pythons this old don't ship with pip by default and installing pip now on py2.6 is non-trivial. Old python compat is also the reason why I use `getopt` instead of `argparse`.

**Comment by webknjaz** [comment]
@tomato42 this looks solvable through trampoline stubs for now, if you want to keep those entry points. But the main runtime code would need to move.

I can make it work so that it'd be both publishable to PyPI and usable from a checkout like you do now. This would postpone the need to figure out completion.

Would you be able to accept such a change? I can send a PR, including updates to deps declarations and automation for publishing to PyPI.

**Comment by tomato42** [comment]
> This change would also make it possible to collect code coverage for things that currently exist in scripts/ and have zero coverage reported, by the way.

that's actually intentional...

**Comment by webknjaz** [comment]
Yeah, I was going to go for something small, anyway. One change would be declaring the deps in `setup.py`, for example.

**Comment by webknjaz** [comment]
Here's how towncrier does trampolines: https://github.com/twisted/towncrier/blob/cd0d2df/bin/towncrier. And here's what we have in ansible: https://github.com/ansible/ansible/tree/989e5833/bin.

*(although, I'd rather get rid of those there since they don't support ancient envs)*

**Comment by tomato42** [comment]
I mean, sure, that would work for `extract.py` or `analysis.py`, but I don't see how that helps with `scripts/`...

**Comment by tomato42** [comment]
if it's no biggie, then please prepare a PR with example

**Comment by webknjaz** [comment]
I've started with a few basic packaging PRs that will be important down the road but don't change anything for you right now. I recommend merging them already, as they should be no-op for how you interact with the project at the moment.

**Comment by webknjaz** [comment]
I'll need to add something on top a bit later, though. After that example, PR (or on top of it).

**Comment by tomato42** [comment]
ok, reviewed them, will wait for CI to pass before merging, and the answers/fixes for the commens

**Comment by webknjaz** [comment]
The move itself is in #994. Technically, the intermediary `_stub.py` wasn't strictly necessary, but it seemed cleaner that way. The symlinks can point to the new location directly.

**Comment by webknjaz** [comment]
@tomato42 talking of publishing from GHA to PyPI, would you like a standalone GHA workflow or one integrated with the CI and running the tests before attempting to publish?

**Comment by tomato42** [comment]
> [@tomato42](https://github.com/tomato42) talking of publishing from GHA to PyPI, would you like a standalone GHA workflow or one integrated with the CI and running the tests before attempting to publish?

not too much of a fan of automated publishing... (and yes I know there are ways for handling secrets, but.... 🫤)

**Comment by webknjaz** [comment]
@tomato42 the modern way relies on OIDC, so you won't have long-living API tokens in secrets. We've also implemented publishing of digital attestations last year, which is basically a standardized substitution for GPG signatures that are no longer supported by PyPI: https://peps.python.org/pep-0740/.

The way I automate this gives you full control over the process and pauses before uploading, allowing you to let it proceed by clicking on a button in the GitHub UI. You don't compromise anything in this workflow.

**Comment by webknjaz** [comment]
I must note that this process allows you to download the about-to-be-published dists for inspection locally. So you can actually check them before they go to PyPI and abort if you don't like something there.

And this doesn't prevent you from using the old school way and uploading from your machine in the future. So you can try it out once and then decide how you like it.

**Comment by webknjaz** [comment]
Here's what I discovered. Note the `inspection` extra is poorly rendered, it should also include numpy and pandas, but mermaid doesn't allow for it, it seems.
```mermaid
graph TD

  subgraph "🗃️ PyPI-sourced deps"
    subgraph "📦 'extraction' extra (optional deps)"
      dpkt["📦 dpkt"]
      pandas["📦 pandas"]
      numpy["📦 numpy"]
    end
    subgraph "📦 'inspection' extra (optional deps, includes numpy too)"
      matplotlib["📦 matplotlib"]
      numpy["📦 numpy"]
      scipy["📦 scipy"]
    end
    subgraph "📦 'compression' extra (optional deps)"
      zstd["📦 zstd"]
    end
  end

  subgraph "🗂️ In-project modules"
    extract("📄 tlsfuzzer.extract")
    analysis("📄 tlsfuzzer.analysis")
    test_tls13_client_certificate_compression("📄 tlsfuzzer._apps.test_tls13_client_certificate_compression")
    stats("📄 tlsfuzzer.utils.stats")
    timing_runner("📄 tlsfuzzer.timing_runner")
    test_bleichenbacher_timing_marvin("📄 tlsfuzzer._apps.test_bleichenbacher_timing_marvin")
    test_lucky13("📄 tlsfuzzer._apps.test_lucky13")
    test_tls13_minerva("📄 tlsfuzzer._apps.test_tls13_minerva")
    test_bleichenbacher_timing_pregenerate("📄 tlsfuzzer._apps.test_bleichenbacher_timing_pregenerate")
  end

%% Import directions:
  dpkt --> extract
  pandas --> analysis
  pandas --> extract
  matplotlib --> analysis
  zstd --> test_tls13_client_certificate_compression
  numpy --> analysis
  numpy --> extract
  numpy --> stats
  scipy --> analysis
  scipy --> stats

  analysis --> timing_runner
  timing_runner --> test_bleichenbacher_timing_marvin
  timing_runner --> test_lucky13
  timing_runner --> test_tls13_minerva
  timing_runner --> test_bleichenbacher_timing_pregenerate

  extract --> timing_runner
  extract --> test_tls13_minerva
  analysis --> stats
```

So the suggested optional dependencies categories (also known as "extras") are as follows:
* `compression`: `zstd`
* `extraction`: `dpkt`, `numpy`, `pandas`
* `inspection`: `matplotlib`, `numpy`, `pandas`, `scipy`

**Comment by tomato42** [comment]
> I must note that this process allows you to download the about-to-be-published dists for inspection locally. So you can actually check them before they go to PyPI and abort if you don't like something there.
> 
> And this doesn't prevent you from using the old school way and uploading from your machine in the future. So you can try it out once and then decide how you like it.

yes, those are good properties, and I didn't know about OIDC

so I have basically just one question about it: how do I decide the versioning of the release? (say, we are at 0.3.2, I just merged a new script to master and want to release a 0.3.3rc1 or a full 0.3.3; how would I go about it?)

**Comment by webknjaz** [comment]
> so I have basically just one question remaining about it: how do I decide the versioning of the release? (say, we are at 0.3.2, I just merged a new script to master and want to release a 0.3.3rc1 or a full 0.3.3; how would I go about it?)

Yeah, I also wanted to raise this question. In my projects, I tend to rely on `setuptools-scm` which essentially computes the version based on the closest Git tag for intermediate commits and the exact tag version number if the current commit is tagged.
The version is computed when building sdists, but past that point, it's saved as a part of the metadata statically.

`setuptools-scm` bumps the last version number component in the discovered tag for the non-tagged commits + adds a dev marker so it's a pre-release. These pre-releases can be published automatically on merges to master. And then, when you want to release a stable version, you'd need to choose what it'll look like, after comparing what's happened since the previous stable release. The exact semantics of version numbers isn't really defined in PEP 440. Some people attempt SemVer or just do whatever they like. It's also popular to use CalVer: https://snarky.ca/why-i-dont-like-semver/. pip migrated to it in 2018, CPython was talking about adopting it in 2026, right after 3.14: https://peps.python.org/pep-2026/ (though, it was rejected at the beginning of this month).

There's all kinds of assumptions the end-users come up with, so it's usually good do document the intended meaning of the version bumps: https://hynek.me/articles/semver-will-not-save-you/.

This also brings a related question on what kind of a trigger you'd like. In my most sophisticated workflows, I really like using the `workflow_dispatch` event @ GHA, which allows me to type in the version into a field on the GH UI and then my workflow pushes that as a tag right after publishing to PyPI.

However, I don't want to overwhelm you with such complexities. So I'd also like to mention other ways:
1. Using `setuptools-scm` but pushing the tag manually. The release workflow would be triggered on the `push` event on GHA. This also allows publishing pre-releases of each merge to master. The pre-releases aren't taken into account by pip's depresolver by default.
2. Not using `setuptools-scm` and hardcoding the version number in the packaging configs. This would involve first making a commit with the version bump and then tagging it. The workflow would run when that tag is pushed. This is prone to human error — the releaser must set the version in two places and not forget to keep them in sync. I've seen some projects also adding another commit right after the release advancing the version to the next patch suffixed with `.dev0`.

I'm not entirely sure how this could integrate with Python 2.6, but I think that it should be possible to make it work on Python 3.13 and use this version in GHA. While 2.6 would not compute the version at all, never hitting `setuptools-scm` in the first place. People getting it from PyPI wouldn't need to build it anyway. And the contributors don't really need to care about the version number in a Git checkout.

As for the version number limitations, the format must conform PEP 440 so that the dists woudld have compliant core packaging metadata. While you could have pre-release versions with dev0/a0/b0/rc0, most projects don't use them. Only those projects that know there are users who're going to actually test these look into publishing pre-releases. The rest just don't care to take on more maintenance burden. One case when I publish alpha releases is when I want to test release automation, which isn't possible from pull requests. So I cut pre-releases to see that it's healthy.
Also, PyPI allows marking versions as yanked, which is not as permanent as removing them but still tells pip to exclude those versions by default. This is a good mechanism for “unpublishing” if the need arises. So I wouldn't overthink it.

CPython uses release candidates to catch final bugs. They have processes in place that forbid certain types of changes past alpha/beta period and release candidates don't get many updates. If a candidate doesn't get bug reports during its cycle, it just gets promoted to a final/stable release, otherwise — a new candidate is made with bugfixes and the cycle repeats. I wouldn't recommend small projects to implement such complex flows, though.

**Comment by webknjaz** [comment]
> hmm... weird... the `test_tls13_client_certificate_compression` can also use `brotli`, not just `zstd` (it does use also `zlib` but that's in the standard library); also it's not so much compression as
> 
> extraction needing `numpy` and `pandas` is a bit surprising... as we do have CI configurations that install `dpkt` but not those two... unfortunately I don't really have time to inspect that now...
> 
> the other thing is that I think the names could be more descriptive of the "why" not "what":
> 
> instead of "compression" using something like "execution" or "extra_deps" or "optional_dependencies" would be better: as those are dependencies that are useful/needed to execute some of the test scripts; also we might want to add `kyber-py` there in the future: once this is solved [GiacomoPope/kyber-py#92](https://github.com/GiacomoPope/kyber-py/issues/92) (and dilithium-py even more far in the future)
> 
> "inspection" is also more "analysis" as it's about performing statistical analysis of the collected timing data... or "statistical-analysis"?

Extras are usually named after the features they enable. Hence, “what”, just high-level.

FWIW, I don't think that debating extras or even adding them should be a blocker for releasing the first version. They can be added later and a new release cut. Or an experimental alpha-release could be made and then released as stable with all the metadata bits too.

I think the main things right now are exposing the scripts via installers + deciding on the release flow triggers. I often organize the jobs to be like "build the dists" -> "test them" -> “release them”, meaning that the existing GHA suite would run while releasing. Some people argue that it's already tested and want a faster workflow without that intermediate part. Once I get this info, I'll give you the instructions for setting up trust on the PyPI side.

Meanwhile, here's what you can configure in the repository settings already. It's not blocked on other decisions:
1. Go to https://github.com/tlsfuzzer/tlsfuzzer/settings/environments
2. Create an environment named `pypi`
3. Add required reviewers (add yourself, or a team if you have those set up), don't prevent self-reviews
4. Optionally, add a wait timer. This allows for having a cooldown period in case you suddenly realize that you want to abort the process.

**Comment by webknjaz** [comment]
> extraction needing `numpy` and `pandas` is a bit surprising... as we do have CI configurations that install `dpkt` but not those two... unfortunately I don't really have time to inspect that now...

They are coming in https://github.com/tlsfuzzer/tlsfuzzer/blob/1147559/build-requirements-analysis.txt via https://github.com/tlsfuzzer/tlsfuzzer/blob/1147559/.github/workflows/ci.yml#L318.

**Comment by webknjaz** [comment]
@tomato42 just an extra confirmation: are we fine with attempts to build the dists under ancient Pythons producing incomplete metadata? Whatever we'll publish to PyPI will work w/o building. And interacting with the repo remains working the same way, bypassing packaging. But newer runtimes will have additional ability to do `pip install [-e]`.

**Comment by tomato42** [comment]
> > extraction needing `numpy` and `pandas` is a bit surprising... as we do have CI configurations that install `dpkt` but not those two... unfortunately I don't really have time to inspect that now...
> 
> They are coming in https://github.com/tlsfuzzer/tlsfuzzer/blob/1147559/build-requirements-analysis.txt via https://github.com/tlsfuzzer/tlsfuzzer/blob/1147559/.github/workflows/ci.yml#L318.

but this is gated behind 
https://github.com/tlsfuzzer/tlsfuzzer/blob/1147559f3f374b1cef7639232e19a9980b4ac4cf/.github/workflows/ci.yml#L314
so the runs with `extract` configuration don't install it...

**Comment by webknjaz** [comment]
> so the runs with `extract` configuration don't install it...

Yes, and in that case the CI just skips the tests that need to import `tlsfuzzer.analysis` (due to missing `scipy` or `matplotlib`, for example: https://github.com/tlsfuzzer/tlsfuzzer/blob/660b60b/tests/test_tlsfuzzer_analysis.py#L19-L32 / https://github.com/tlsfuzzer/tlsfuzzer/blob/660b60b/tests/test_tlsfuzzer_timing_runner.py#L140-L147.

**Comment by webknjaz** [comment]
The skipped tests aren't printed out because the test runner is not invoked with verbosity: https://github.com/tlsfuzzer/tlsfuzzer/actions/runs/13634077693/job/38108412799#step:27:10. But I'm pretty sure those would be the skipped entries caused by the missing deps.

**Comment by tomato42** [comment]
sorry for the delay

so, in order:
https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2685560458

computing the version is ok; yes, having an easy way to push a "dev" tarball is quite a good idea for tlsfuzzer, maybe even after every merge

nah... I don't think the issues with SemVer convince me to go for CalVer; I'm using SemVer in both dependent projects and I'm quite happy with it

> However, I don't want to overwhelm you with such complexities. So I'd also like to mention other ways:

hmm, the issue is that we at the very least have version in `setup.py` and `docs/source/conf.py`, also I like having a `<package>.__version__` field that's hardcoded in the released code...

> I'm not entirely sure how this could integrate with Python 2.6, but I think that it should be possible to make it work on Python 3.13 and use this version in GHA. While 2.6 would not compute the version at all, never hitting setuptools-scm in the first place. People getting it from PyPI wouldn't need to build it anyway. And the contributors don't really need to care about the version number in a Git checkout.

py2.6 is so old and unlikely to be used in practice that we might just say "checkout the code, use `PYTHONPATH=.` for running it"; that's how we run in it practice anyway...

> While you could have pre-release versions with dev0/a0/b0/rc0, most projects don't use them

I do use them extensively: I frequently have need to use pre-release code but don't consider it good enough for others to use. 😅 

https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2685583362

> FWIW, I don't think that debating extras or even adding them should be a blocker for releasing the first version. 

well, I think this is solved already 🙂 

> meaning that the existing GHA suite would run while releasing.

hmm... well, depends on how we can handle the "putting version in at least 3 places" part; if we can automate this, then doing full CI run before pushing the release to PyPI is the only reasonable approach in my book; if it turns out that I have to bump the version in the files myself (not a blocker for me at all, I've been doing it for like a decade in tlslite-ng already), and push that together with new tag, then yes, CI won't be run; and that's fine too: I run it before *and* after merges anyway...

> Create an environment named `pypi`

done

https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2688490852

yes, what is confusing for me is that the extract.py script can't run without them but we supposedly test it without them...
(https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2694858436) I'm starting to suspect that those "extract" runs on CI are just burning minutes... 😅 

https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2690568949

> [@tomato42](https://github.com/tomato42) just an extra confirmation: are we fine with attempts to build the dists under ancient Pythons producing incomplete metadata? Whatever we'll publish to PyPI will work w/o building. And interacting with the repo remains working the same way, bypassing packaging. But newer runtimes will have additional ability to do `pip install [-e]`.

what do you mean by "incomplete metadata"? and as I wrote above, yes, I'm OK with the package being uninstallable on py2.6... not as sure about py2.7...

**Comment by webknjaz** [comment]
> sorry for the delay [@webknjaz](https://github.com/webknjaz)
> 
> so, in order: [#903 (comment)](https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2685560458)
> 
> computing the version is ok; yes, having an easy way to push a "dev" tarball is quite a good idea for tlsfuzzer, maybe even after every merge
>

Note that this is done on the packaging level and written to metadata, not to `.py` files (by default).

> nah... I don't think the issues with SemVer convince me to go for CalVer; I'm using SemVer in both dependent projects and I'm quite happy with it
>

Sure, packaging doesn't really restrict choosing whatever you like.

> > However, I don't want to overwhelm you with such complexities. So I'd also like to mention other ways:
> 
> hmm, the issue is that we at the very least have version in `setup.py` and `docs/source/conf.py`, also I like having a `<package>.__version__` field that's hardcoded in the released code...
>

The version is computed for inclusion into static core packaging metadata. After that, it can be read from installed metadata programmatically. This way, there's a single source of version — Git.
The version in `__init__.py` wouldn't be hardcoded. It would be read from another source. `setuptools-scm` allow additionally writing it to a file if you want, though.
Using `setuptools-scm` would remove the version number from source code in Git and move it to Git tags. When integrated, `setup.py` doesn't need to hardcode it as it's computed by the setuptools plugin. Sphinx can also source it from installed package metadata, when the pkg is installed, which is what is currently recommended.
But if you really want to, a Python module can be generated when packaging and then used for getting the version into runtime. It's just one extra setting in the config.

Regarding `__version__`, it's been controversial lately. We had a lot of heated arguments over @ PyPUG like https://discuss.python.org/t/please-make-package-version-go-away/58501/85 / https://github.com/pypa/packaging.python.org/pull/1276 / https://github.com/pypa/packaging.python.org/pull/1273 / https://github.com/pypa/packaging.python.org/pull/1578 / https://github.com/pypa/packaging.python.org/pull/1580 / https://github.com/pypa/packaging.python.org/pull/1607.
We ended up with these docs: https://packaging.python.org/en/latest/discussions/single-source-version/ / https://packaging.python.org/en/latest/discussions/versioning/#runtime-version-access.

The bottom-line is that you either hardcode your version in `__version__` and make both packaging and import time access read from that, or you don't. When you source the version during packaging, you then access it through standard methods for reading that metadata and can populate `__version__` from the same metadata. `setuptools-scm` can also let you have hybrid approach where it'd put the version from Git tag into the metadata but also generate a Python module where you can import it from and assign where you like.

When building Sphinx docs from Git, you'll need to make sure to adjust the fetch depth because various CI/build platforms only grab the last commit and no tags by default. This is not a big deal, just something to be aware of.

> > I'm not entirely sure how this could integrate with Python 2.6, but I think that it should be possible to make it work on Python 3.13 and use this version in GHA. While 2.6 would not compute the version at all, never hitting setuptools-scm in the first place. People getting it from PyPI wouldn't need to build it anyway. And the contributors don't really need to care about the version number in a Git checkout.
> 
> py2.6 is so old and unlikely to be used in practice that we might just say "checkout the code, use `PYTHONPATH=.` for running it"; that's how we run in it practice anyway...

That's what I assumed. Just wanted to be sure.

> > While you could have pre-release versions with dev0/a0/b0/rc0, most projects don't use them
> 
> I do use them extensively: I frequently have need to use pre-release code but don't consider it good enough for others to use. 😅

This might be fine sometimes. It's up to you. Although, it's a good idea to stop recommending `--pre` in `pip install` which influences every single dist in the dependency tree. Instead, doing something like `pip install 'pkg >= 1.2.3a0'` enables `--pre` but just for that single dist, granularly, which limits the damage.

> [#903 (comment)](https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2685583362)
> 
> > FWIW, I don't think that debating extras or even adding them should be a blocker for releasing the first version.
> 
> well, I think this is solved already 🙂

Yes. But if you'll want to add more, you can still tell people to install an extra dep manually even if extras don't fit the need…

> > meaning that the existing GHA suite would run while releasing.
> 
> hmm... well, depends on how we can handle the "putting version in at least 3 places" part; if we can automate this, then doing full CI run before pushing the release to PyPI is the only reasonable approach in my book; if it turns out that I have to bump the version in the files myself (not a blocker for me at all, I've been doing it for like a decade in tlslite-ng already), and push that together with new tag, then yes, CI won't be run; and that's fine too: I run it before _and_ after merges anyway...
>

I think the most popular way (with balanced complexity) I've seen is that people would push a Git tag w/o the need to bump hardcoded versions in the source code, in which case CI can still run and abort the release.

Anyway, I'm targeting putting the version into a single place — Git tag. It's just that I'm asking if you want to have `git tag -s v1.2.3; git push origin v1.2.3` as a trigger which ends up publishing the release, or you can have a UI element on GitHub where you'd type in the version number also, just once, and that would end up publishing to PyPI and pushing the tag for you. I'm personally a fan of the latter, as it removes human error from more places.

> > Create an environment named `pypi`
> 
> done
>

Did you set up required reviewers too?

> [#903 (comment)](https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2688490852)
> 
> yes, what is confusing for me is that the extract.py script can't run without them but we supposedly test it without them... ([#903 (comment)](https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2694858436)) I'm starting to suspect that those "extract" runs on CI are just burning minutes... 😅

Yep. I usually use Codecov, not Coveralls. And it allows applying flags to coverage uploads. So with those flags I can see which lines were covered in which CI runs (or under which Python / OS versions etc.)

> [#903 (comment)](https://github.com/tlsfuzzer/tlsfuzzer/issues/903#issuecomment-2690568949)
> 
> > [@tomato42](https://github.com/tomato42) just an extra confirmation: are we fine with attempts to build the dists under ancient Pythons producing incomplete metadata? Whatever we'll publish to PyPI will work w/o building. And interacting with the repo remains working the same way, bypassing packaging. But newer runtimes will have additional ability to do `pip install [-e]`.
> 
> what do you mean by "incomplete metadata"? and as I wrote above, yes, I'm OK with the package being uninstallable on py2.6... not as sure about py2.7...

I'm not talking about being installable specifically. I was considering using newer declarative configs that might not be picked up on ancient envs. This wouldn't break the existing runs from Git checkout.
I'm mostly thinking of the installable scripts that I haven't yet added to the packaging configuration. Things that appear in `venv/bin/` as you `pip install` the project.
If I use newer configs for some things, then the CI/CD that publishes to PyPI would embed it into the dists correctly, and `pip install` from PyPI under any runtime would get this metadata, and pip would place those scripts where they should be.
But on older systems, when building from sdist (the tarball, not the zip file/wheel) or from Git, this config might not be picked up and so the scripts wouldn't be installed. But it would still produce packaging metadata with the runtime deps, for example. This is the case with doing the discouraged `python setup.py sdist bdist_wheel install` invocations. So it wouldn't really affect real users.

I guess I can just try it out, and you could test if it works for you where it's important.

**Comment by tomato42** [comment]
> But if you really want to, a Python module can be generated when packaging and then used for getting the version into runtime. It's just one extra setting in the config.

I don't think I like the dependency on `setuptools-scm`, so delegating it to a single file that's then imported all over is the better approach... (so that it works with both checkout version of the tools and in install from pypi)

> Regarding `__version__`, it's been controversial lately

maybe, but a static one, that doesn't depend on external things, is kind of necessary if we're talking about compatibility back to py2.6

so having to manually update it is a little price to pay (and yes, the glorious future would be nice to arrive, but it hasn't been distributed evenly 😉 )

> Although, it's a good idea to stop recommending --pre in pip install which influences every single dist in the dependency tree. Instead, doing something like pip install 'pkg >= 1.2.3a0' enables --pre but just for that single dist, granularly, which limits the damage.

agree, it's suboptimal; but then it's not like `tlslite-ng` is a widely used package, and the pre-releases are still in good shape, so it's not like using them, and affecting other packages is likely to cause problems (not with the level of API stability that I go for)

> Yes. But if you'll want to add more, you can still tell people to install an extra dep manually even if extras don't fit the need…

👍  not the first time I used extras in a project

> Anyway, I'm targeting putting the version into a single place — Git tag. It's just that I'm asking if you want to have git tag -s v1.2.3; git push origin v1.2.3 as a trigger which ends up publishing the release, or you can have a UI element on GitHub where you'd type in the version number also, just once, and that would end up publishing to PyPI and pushing the tag for you. I'm personally a fan of the latter, as it removes human error from more places.

not going to lie... the latter option is quite tempting, but I wonder how likely it is to _continue_ to work... Microsoft keeps on messing with actions and available workflows all the time (deprecating stuff that works for arbitrary reasons), and frankly, it's _super_ irritating, so I personally prefer a bit more involved workflow, but one that I understand fully, so I know how to fix it _when_ (not if) it breaks...

> Did you set up required reviewers too?

yes, though set for myself only for now

> Yep. I usually use Codecov, not Coveralls. And it allows applying flags to coverage uploads. So with those flags I can see which lines were covered in which CI runs (or under which Python / OS versions etc.)

coveralls does that too, if a line wasn't executed under all environments then it will show which ones have executed it (when it actually works... the whole service has been pretty flaky lately...)

> But on older systems, when building from sdist (the tarball, not the zip file/wheel) or from Git, this config might not be picked up and so the scripts wouldn't be installed. 

I get it, but I'm afraid that not having that working with py2.7 is a bit of a show-stopper... (_maybe_ we might swap py2.7 for py3.6, but I'm afraid that keeping it working on py3.6 may end up being more complicated than keeping it working on py2.7...)

**Comment by webknjaz** [comment]
> > But if you really want to, a Python module can be generated when packaging and then used for getting the version into runtime. It's just one extra setting in the config.
> 
> I don't think I like the dependency on `setuptools-scm`, so delegating it to a single file that's then imported all over is the better approach... (so that it works with both checkout version of the tools and in install from pypi)

That would be a packaging-only dependency. Without it, it's re-hardcoding the version string before and after each release. It would not be picked up by older envs like Python 2.6 which could fall back to something like an "unversioned" string.

> > Regarding `__version__`, it's been controversial lately
> 
> maybe, but a static one, that doesn't depend on external things, is kind of necessary if we're talking about compatibility back to py2.6
> 
> so having to manually update it is a little price to pay (and yes, the glorious future would be nice to arrive, but it hasn't been distributed evenly 😉 )

Since we were talking about having Python 2.6 work as before, out of the repo, the versioning information wouldn't be important in that env, so the integration could just be skipped. There's already a dependency on `setuptools` in PEP 517-aware envs — it's going to be `pip install`ed during packaging. This would add another even smaller dep. And ancient envs that don't know what PEP 517 is would not be affected by this in any way.
Additionally, since building dists for publishing to PyPI would be happening under something modern like Python 3.13, this integration would work. And the PyPI-shipped dists would embed the generated version statically. So when the end-users on Python 2.6 would install those wheels from PyPI, they'll have this information and no computation would be happening in their envs.

> > Although, it's a good idea to stop recommending --pre in pip install which influences every single dist in the dependency tree. Instead, doing something like pip install 'pkg >= 1.2.3a0' enables --pre but just for that single dist, granularly, which limits the damage.
> 
> agree, it's suboptimal; but then it's not like `tlslite-ng` is a widely used package, and the pre-releases are still in good shape, so it's not like using them, and affecting other packages is likely to cause problems (not with the level of API stability that I go for)

Yeah, but I'm talking about any other things the end-users might be installing in the same virtualenv. The harm exists outside your dependency. But suggesting a granular command would address that easily.


> > Anyway, I'm targeting putting the version into a single place — Git tag. It's just that I'm asking if you want to have git tag -s v1.2.3; git push origin v1.2.3 as a trigger which ends up publishing the release, or you can have a UI element on GitHub where you'd type in the version number also, just once, and that would end up publishing to PyPI and pushing the tag for you. I'm personally a fan of the latter, as it removes human error from more places.
> 
> not going to lie... the latter option is quite tempting, but I wonder how likely it is to _continue_ to work... Microsoft keeps on messing with actions and available workflows all the time (deprecating stuff that works for arbitrary reasons), and frankly, it's _super_ irritating, so I personally prefer a bit more involved workflow, but one that I understand fully, so I know how to fix it _when_ (not if) it breaks...

The important things remain working. There's been a few bumps along the way, though, which is why I'm working on my own framework for that. Also because I maintain a lot of things, and so I need something that I can reuse in tens of repos, while reducing the maintenance cost. The `pypi-publish` thing was probably the very first GitHub Action I made for public consumption and now it's the PyPA-blessed go-to mechanism.

What I have is not that complicated, though. But it may be a lot of lines.

> > But on older systems, when building from sdist (the tarball, not the zip file/wheel) or from Git, this config might not be picked up and so the scripts wouldn't be installed.
> 
> I get it, but I'm afraid that not having that working with py2.7 is a bit of a show-stopper... (_maybe_ we might swap py2.7 for py3.6, but I'm afraid that keeping it working on py3.6 may end up being more complicated than keeping it working on py2.7...)

If you're not opposed to requiring that there's a modern enough pip version, it should work. And anyway, the end-users would not be building it but getting wheels from PyPI. So unless they force building from sdist, it'll never affect them.

**Comment by webknjaz** [comment]
> > > > Regarding `__version__`, it's been controversial lately
> > > 
> > > 
> > > maybe, but a static one, that doesn't depend on external things, is kind of necessary if we're talking about compatibility back to py2.6
> > > so having to manually update it is a little price to pay (and yes, the glorious future would be nice to arrive, but it hasn't been distributed evenly 😉 )
> > 
> > 
> > Since we were talking about having Python 2.6 work as before, out of the repo, the versioning information wouldn't be important in that env, so the integration could just be skipped. There's already a dependency on `setuptools` in PEP 517-aware envs — it's going to be `pip install`ed during packaging. This would add another even smaller dep. And ancient envs that don't know what PEP 517 is would not be affected by this in any way. Additionally, since building dists for publishing to PyPI would be happening under something modern like Python 3.13, this integration would work. And the PyPI-shipped dists would embed the generated version statically. So when the end-users on Python 2.6 would install those wheels from PyPI, they'll have this information and no computation would be happening in their envs.
> 
> Well, the thing is that I would like to have `print("running with version: {0}".format(tlsfuzzer.__version__))` in the scripts, in the summary; generally working even when running from a checkout (doesn't have to be exact or precise, but it should return something at least like "the most recent tag in history")

Under Python 2.6? Also, this seems like a feature creep. This issue is about publishing to PyPI. Current code doesn't output this and the metadata is set to `0.0.1`. Let's postpone this until it's packaged (or at least that is mostly figured out). I feel like we're deviating into several standalone contexts here, and I'd like to focus on packaging right now.

> > > > Anyway, I'm targeting putting the version into a single place — Git tag. It's just that I'm asking if you want to have git tag -s v1.2.3; git push origin v1.2.3 as a trigger which ends up publishing the release, or you can have a UI element on GitHub where you'd type in the version number also, just once, and that would end up publishing to PyPI and pushing the tag for you. I'm personally a fan of the latter, as it removes human error from more places.
> > > 
> > > 
> > > not going to lie... the latter option is quite tempting, but I wonder how likely it is to _continue_ to work... Microsoft keeps on messing with actions and available workflows all the time (deprecating stuff that works for arbitrary reasons), and frankly, it's _super_ irritating, so I personally prefer a bit more involved workflow, but one that I understand fully, so I know how to fix it _when_ (not if) it breaks...
> > 
> > 
> > The important things remain working. There's been a few bumps along the way, though, which is why I'm working on my own framework for that. Also because I maintain a lot of things, and so I need something that I can reuse in tens of repos, while reducing the maintenance cost. The `pypi-publish` thing was probably the very first GitHub Action I made for public consumption and now it's the PyPA-blessed go-to mechanism.
> > What I have is not that complicated, though. But it may be a lot of lines.
> 
> not against, especially if it solves the issue of `tlsfuzzer.__version` automatically (by like updating the value, and pushing the commit to `master` directly, together with the tag?)
>

The thing is that the hardcoded value is never committed to Git via its content, only the tag is pushed. Unless you'll want to have some other things like a generated change log, that is. Whenever there's Towncrier, I do indeed also push the commit.

I think we've gotten into abstract discussions too much already. It's probably better if I just show, and you play with corresponding PRs.

**Comment by webknjaz** [comment]
Submitted #999 for a better demo.

**Comment by tomato42** [comment]
First of all, sorry for the late reply, I'm swamped with work because of internal deadlines and IETF 122 happening at the same time.

> > > > > Regarding `__version__`, it's been controversial lately
> > > > 
> > > > 
> > > > maybe, but a static one, that doesn't depend on external things, is kind of necessary if we're talking about compatibility back to py2.6
> > > > so having to manually update it is a little price to pay (and yes, the glorious future would be nice to arrive, but it hasn't been distributed evenly 😉 )
> > > 
> > > 
> > > Since we were talking about having Python 2.6 work as before, out of the repo, the versioning information wouldn't be important in that env, so the integration could just be skipped. There's already a dependency on `setuptools` in PEP 517-aware envs — it's going to be `pip install`ed during packaging. This would add another even smaller dep. And ancient envs that don't know what PEP 517 is would not be affected by this in any way. Additionally, since building dists for publishing to PyPI would be happening under something modern like Python 3.13, this integration would work. And the PyPI-shipped dists would embed the generated version statically. So when the end-users on Python 2.6 would install those wheels from PyPI, they'll have this information and no computation would be happening in their envs.
> > 
> > 
> > Well, the thing is that I would like to have `print("running with version: {0}".format(tlsfuzzer.__version__))` in the scripts, in the summary; generally working even when running from a checkout (doesn't have to be exact or precise, but it should return something at least like "the most recent tag in history")
> 
> Under Python 2.6? Also, this seems like a feature creep. This issue is about publishing to PyPI. Current code doesn't output this and the metadata is set to `0.0.1`. Let's postpone this until it's packaged (or at least that is mostly figured out). I feel like we're deviating into several standalone contexts here, and I'd like to focus on packaging right now.

the reason why I wasn't publishing versions was because you were supposed to get it from git, and thus have a specific commit id to reference it; once we detach code from git, we need to identify which version it is

> > > > > Anyway, I'm targeting putting the version into a single place — Git tag. It's just that I'm asking if you want to have git tag -s v1.2.3; git push origin v1.2.3 as a trigger which ends up publishing the release, or you can have a UI element on GitHub where you'd type in the version number also, just once, and that would end up publishing to PyPI and pushing the tag for you. I'm personally a fan of the latter, as it removes human error from more places.
> > > > 
> > > > 
> > > > not going to lie... the latter option is quite tempting, but I wonder how likely it is to _continue_ to work... Microsoft keeps on messing with actions and available workflows all the time (deprecating stuff that works for arbitrary reasons), and frankly, it's _super_ irritating, so I personally prefer a bit more involved workflow, but one that I understand fully, so I know how to fix it _when_ (not if) it breaks...
> > > 
> > > 
> > > The important things remain working. There's been a few bumps along the way, though, which is why I'm working on my own framework for that. Also because I maintain a lot of things, and so I need something that I can reuse in tens of repos, while reducing the maintenance cost. The `pypi-publish` thing was probably the very first GitHub Action I made for public consumption and now it's the PyPA-blessed go-to mechanism.
> > > What I have is not that complicated, though. But it may be a lot of lines.
> > 
> > 
> > not against, especially if it solves the issue of `tlsfuzzer.__version` automatically (by like updating the value, and pushing the commit to `master` directly, together with the tag?)
> 
> The thing is that the hardcoded value is never committed to Git via its content, only the tag is pushed. Unless you'll want to have some other things like a generated change log, that is. Whenever there's Towncrier, I do indeed also push the commit.
> 
> I think we've gotten into abstract discussions too much already. It's probably better if I just show, and you play with corresponding PRs.

"Towncrier"?

we're talking about abstracts because I'm not familiar with python packaging, I know very little so I don't know which things have what effects, and I really don't have time to research them, sorry; I can only say how the end solution is supposed to behave
